### PR TITLE
new feature: Warehouse

### PR DESCRIPTION
**Pull request by MineYuanlu**
## Description
a new feature: warehosue
it can auto sort all the chest which you want: #518 

## screenshots / videos
图片

[video](https://youtu.be/3KW7rgOgZc0)

### CODE REVIEW

**Comment by MineYuanlu** [comment]
@Alexander01998  I really want your opinion. I don't know how this idea is.

File: gradle.properties [hunk 1]
```diff
@@ -4,15 +4,15 @@ org.gradle.jvmargs=-Xmx1G
 # Fabric Properties
 # check these on https://modmuss50.me/fabric.html and
 # https://www.curseforge.com/minecraft/mc-mods/fabric-api
-minecraft_version=1.18.1
-yarn_mappings=1.18.1+build.2
+minecraft_version=1.18
+yarn_mappings=1.18+build.1
```
Comment (KTibow) [inline gradle.properties:8]:
```suggestion
minecraft_version=1.18.1
yarn_mappings=1.18.1+build.2
```

File: gradle.properties [hunk 2]
```diff
@@ -4,15 +4,15 @@ org.gradle.jvmargs=-Xmx1G
...
+minecraft_version=1.18
+yarn_mappings=1.18+build.1
 loader_version=0.12.12
 
 #Fabric api
 fabric_version=0.44.0+1.18
 
 # Mod Properties
-mod_version = v7.19-MC1.18.1
+mod_version = v7.19-MC1.18
```
Comment (KTibow) [inline gradle.properties:15]:
```suggestion
mod_version = v7.19-MC1.18.1
```

File: src/main/java/net/wurstclient/WurstClient.java [hunk 3]
```diff
@@ -58,7 +58,7 @@
 	public static final IMinecraftClient IMC = (IMinecraftClient)MC;
 	
 	public static final String VERSION = "7.19";
-	public static final String MC_VERSION = "1.18.1";
+	public static final String MC_VERSION = "1.18";
```
Comment (KTibow) [inline src/main/java/net/wurstclient/WurstClient.java:61]:
```suggestion
	public static final String MC_VERSION = "1.18.1";
```

File: src/main/java/net/wurstclient/command/CmdList.java [hunk 4]
```diff
@@ -64,6 +64,7 @@
 	public final UnbindCmd unbindCmd = new UnbindCmd();
 	public final VClipCmd vClipCmd = new VClipCmd();
 	public final ViewNbtCmd viewNbtCmd = new ViewNbtCmd();
+	public final WarehouseCmd warehouseCmd=new WarehouseCmd();
```
Comment (KTibow) [inline src/main/java/net/wurstclient/command/CmdList.java:67]:
```suggestion
	public final WarehouseCmd warehouseCmd = new WarehouseCmd();

```

File: src/main/java/net/wurstclient/mixin/ClientPlayNetworkHandlerMixin.java [hunk 5]
```diff
@@ -15,13 +15,21 @@
 import net.minecraft.client.network.ClientPlayNetworkHandler;
 import net.minecraft.network.Packet;
 import net.minecraft.network.listener.ClientPlayPacketListener;
+import net.minecraft.network.packet.s2c.play.InventoryS2CPacket;
+import net.wurstclient.WurstClient;
+import net.wurstclient.commands.WarehouseCmd;
 import net.wurstclient.event.EventManager;
 import net.wurstclient.events.PacketOutputListener.PacketOutputEvent;
 
+
```
Comment (KTibow) [inline src/main/java/net/wurstclient/mixin/ClientPlayNetworkHandlerMixin.java:24]:
```suggestion
```

File: src/main/java/net/wurstclient/mixin/ContainerScreen54Mixin.java [hunk 6]
```diff
@@ -102,7 +107,6 @@ private void shiftClickSlots(int from, int to, int mode)
 			waitForDelay();
 			if(this.mode != mode || client.currentScreen == null)
 				break;
-			
 			onMouseClick(slot, slot.id, 0, SlotActionType.QUICK_MOVE);
```
Comment (KTibow) [inline src/main/java/net/wurstclient/mixin/ContainerScreen54Mixin.java:110]:
```suggestion

			onMouseClick(slot, slot.id, 0, SlotActionType.QUICK_MOVE);
```

File: src/main/java/net/wurstclient/mixin/ShulkerBoxScreenMixin.java [hunk 7]
```diff
@@ -31,6 +32,8 @@
 	
 	private final AutoStealHack autoSteal =
 		WurstClient.INSTANCE.getHax().autoStealHack;
+	private final WarehouseCmd warehouse =
+			WurstClient.INSTANCE.getCmds().warehouseCmd;
```
Comment (KTibow) [inline src/main/java/net/wurstclient/mixin/ShulkerBoxScreenMixin.java:36]:
```suggestion
		WurstClient.INSTANCE.getCmds().warehouseCmd;
```

File: src/main/java/net/wurstclient/mixin/ShulkerBoxScreenMixin.java [hunk 8]
```diff
@@ -48,6 +51,8 @@ protected void init()
 		if(!WurstClient.INSTANCE.isEnabled())
 			return;
 		
+		if (warehouse.callbackOpenWindow(handler.syncId,rows*9)) return;
```
Comment (KTibow) [inline src/main/java/net/wurstclient/mixin/ShulkerBoxScreenMixin.java:54]:
```suggestion
		if (warehouse.callbackOpenWindow(handler.syncId, rows*9)) return;

```

**Review by KTibow — CHANGES_REQUESTED** [review]
This looks mostly good. It looks like you accidentally set the version to 1.18 instead of 1.18.1, so I fixed that. Also I had a couple of style tweaks.

**Comment by MineYuanlu** [comment]
> This looks mostly good. It looks like you accidentally set the version to 1.18 instead of 1.18.1, so I fixed that. Also I had a couple of style tweaks.

I am very grateful to you for helping me solve these problems! 😁 😁

**Comment by KTibow** [comment]
Now, given that Wurst tends to be more inactive, I think it would be a good idea if you could upload a build of Wurst with your feature publicly, while you wait for them to merge your pull request.

**Comment by MineYuanlu** [comment]
> Now, given that Wurst tends to be more inactive, I think it would be a good idea if you could upload a build of Wurst with your feature publicly, while you wait for them to merge your pull request.

OK, I'll upload it after I fix all the known problems.

**Comment by MineYuanlu** [comment]
I think there are problems with some operations on git branch in the past. I will restart the pull request after finishing the code. The code has moved to a new branch: [new branch - warehouse](https://github.com/MineYuanlu/Wurst7/tree/warehouse)

**Comment by KTibow** [comment]
That branch looks good. Make sure to let me know when you're finished by putting the text "@KTibow" in your pull request.

**Comment by MineYuanlu** [comment]
@KTibow 


Look here: https://github.com/MineYuanlu/Wurst7/releases  
I have a little time today. I fixed some bugs and sorted out some things

**Comment by KTibow** [comment]
I'll look at it soon. Thanks.

**Comment by KTibow** [comment]
@MineYuanlu Can you upload a video of you configuring Warehouse, without any previous configuration? I can't figure out how to get started.

**Comment by KTibow** [comment]
Also I would recommend changing the description for TradeFill to:
```
Automatically inserts emeralds into the villager trade menu.
```

**Comment by MineYuanlu** [comment]
> @MineYuanlu Can you upload a video of you configuring Warehouse, without any previous configuration? I can't figure out how to get started.

[warehouse wiki](https://github.com/MineYuanlu/Wurst7/wiki/Warehouse)

I made a simple wiki page.
My English is not very good. Most of them use machine translation and may not be able to express my meaning clearly.

**Comment by KTibow** [comment]
How do you specify the item list for an output chest?

**Comment by MineYuanlu** [comment]
> How do you specify the item list for an output chest?

Oh, sorry, I forgot to say. Put items in the box, and then enter the sign mode. Right click the box to mark.

Now Beijing time (UTC + 8) has just entered a new year. Happy New Year!

**Comment by KTibow** [comment]
I would recommend changing
```
.warehouse sign <type> [w] [io] [a] Enable container tagging
```
to
```
.warehouse markmode <type> [priority] [maxAmountMode] [maxAmount] Go into marking mode
```

**Comment by KTibow** [comment]
Also, maybe default maxAmountMode and amount to item and -27. Other than that, I got this to work, which seems really cool.

**Comment by MineYuanlu** [comment]
> I would recommend changing
> 
> ```
> .warehouse sign <type> [w] [io] [a] Enable container tagging
> ```
> 
> to
> 
> ```
> .warehouse markmode <type> [priority] [maxAmountMode] [maxAmount] Go into marking mode
> ```

well! I have modified the content of the wiki, and the tips in the game will be modified in the next version



> Also, maybe default maxAmountMode and amount to item and -27. Other than that, I got this to work, which seems really cool.

At present, the default `maxAmountMode` is `item` and `maxAmount` is `0`. I think `maxAmount` in `storage` mode should be changed to `-54`(considering the double-chest, it should not be `-27`), and others should remain unchanged. What's your opinion?

**Comment by KTibow** [comment]
> well! I have modified the content of the wiki, and the tips in the game will be modified in the next version

Sounds good.

> think maxAmount in storage mode should be changed to -54(considering the double-chest, it should not be -27), and others should remain unchanged. What's your opinion?

Well, what would happen if you try to use `-54` on a single chest? Would anything break?

**Comment by MineYuanlu** [comment]
> Well, what would happen if you try to use `-54` on a single chest? Would anything break?

No, the program will detect whether it can be stored. This mechanism is a little complex, but I think there is no problem. I have tested it for a long time.

Then, can we exchange other contact information? I don't think it's good to communicate here.

**Comment by KTibow** [comment]
> No, the program will detect whether it can be stored. This mechanism is a little complex, but I think there is no problem. I have tested it for a long time.
Then, can we exchange other contact information? I don't think it's good to communicate here.

Alright, that sounds good. I'm KTibow#3960 on Discord, feel free to send a friend request.

**Pull request closed without merging**